src/cryptonote_core/service_node_list.cpp

```diff
@@ -256,34 +265,87 @@ namespace service_nodes
     if (expiration_timestamp < block_timestamp)
       return false;
 
+    if (service_node_addresses.size() > MAX_NUMBER_OF_SHAREHOLDERS)
+      return false;
+
+    key = service_node_key;
+
+    info.last_reward_block_height = block_height;
+    info.last_reward_transaction_index = index;
+    info.addresses = service_node_addresses;
+    info.shares = service_node_shares;
+    info.total_contributions = 0;
+    info.staking_requirement = m_blockchain.get_staking_requirement(block_height);
+
+    return true;
+  }
+
+  void service_node_list::process_registration_tx(const cryptonote::transaction& tx, uint64_t block_timestamp, uint64_t block_height, int index)
+  {
+    crypto::public_key key;
+    service_node_info info;
+    if (!is_registration_tx(tx, block_timestamp, block_height, index, key, info))
+      return;
+
+    auto iter = m_service_nodes_infos.find(key);
+    if (iter != m_service_nodes_infos.end())
+      return;
+
+    MGINFO("New service node registered: " << key);
+
+    m_rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_new(block_height, key)));
+    m_service_nodes_infos[key] = info;
+  }
+
+  void service_node_list::process_contribution_tx(const cryptonote::transaction& tx, uint64_t block_height, int index)
```

Comment:
I don't see any way for this function to mark a processed tx as invalid (if, for example, it's missing fields or too low of a contribution).  I presume that happens elsewhere?

Reply:
We decided that for this release registration txs (and also now contribution txs) will be regular txs with some other stuff in tx extra.

If this fails then the tx is just a regular tx.

Reply:
hmm...perhaps there should at least be some sanity/validity check on the incoming tx side of things, so people don't accidentally lock money that won't be useful.

Reply:
Yeah that's probably a good idea. I'll take a look at it